src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/rc_command_request.cc

```diff
@@ -0,0 +1,282 @@
+  LOG4CXX_AUTO_TRACE(logger_);
+  std::string info;
+  if (rc_rpc_plugin::kDisallowed == access) {
+    info = disallowed_info_.empty()
+               ? "The RPC is disallowed by vehicle settings"
+               : disallowed_info_;
+  } else {
+    return;
+  }
+  LOG4CXX_ERROR(logger_, info);
+  SetResourceState(ModuleType(), ResourceState::FREE);
+  SendResponse(false, mobile_apis::Result::DISALLOWED, info.c_str());
+}
+
+void RCCommandRequest::Run() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  app_mngr::ApplicationSharedPtr app =
+      application_manager_.application(CommandRequestImpl::connection_key());
+
+  if (!IsInterfaceAvailable(app_mngr::HmiInterfaces::HMI_INTERFACE_RC)) {
+    LOG4CXX_WARN(logger_, "HMI interface RC is not available");
+    SetResourceState(ModuleType(), ResourceState::FREE);
+    SendResponse(false,
+                 mobile_apis::Result::UNSUPPORTED_RESOURCE,
+                 "Remote control is not supported by system");
+    return;
+  }
+  LOG4CXX_TRACE(logger_, "RC interface is available!");
+  if (!policy_handler_.CheckHMIType(
+          app->policy_app_id(),
+          mobile_apis::AppHMIType::eType::REMOTE_CONTROL,
+          app->app_types())) {
+    LOG4CXX_WARN(logger_, "Application has no remote control functions");
+    SetResourceState(ModuleType(), ResourceState::FREE);
+    SendResponse(false, mobile_apis::Result::DISALLOWED, "");
+    return;
+  }
+  if (!resource_allocation_manager_.is_rc_enabled()) {
+    LOG4CXX_WARN(logger_, "Remote control is denied by user");
+    SetResourceState(ModuleType(), ResourceState::FREE);
+    SendResponse(false,
+                 mobile_apis::Result::UNSUPPORTED_RESOURCE,
+                 "Remote control is denied by user");
+    return;
+  }
+
+  if (CheckDriverConsent()) {
+    if (AcquireResources()) {
+      Execute();  // run child's logic
+    }
+    // If resource is not aqcuired, AcquireResources method will either
+    // send response to mobile or
+    // send additional request to HMI to ask driver consent
+  }
+}
+
+bool RCCommandRequest::AcquireResources() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  const std::string module_type = ModuleType();
+
+  if (!IsResourceFree(module_type)) {
+    LOG4CXX_WARN(logger_, "Resource is busy.");
+    SetResourceState(ModuleType(), ResourceState::FREE);
+    SendResponse(false, mobile_apis::Result::IN_USE, "");
+    return false;
+  }
+
+  AcquireResult::eType acquire_result = AcquireResource(message_);
+  switch (acquire_result) {
+    case AcquireResult::ALLOWED: {
+      SetResourceState(module_type, ResourceState::BUSY);
+      return true;
+    }
+    case AcquireResult::IN_USE: {
+      SetResourceState(ModuleType(), ResourceState::FREE);
+      SendResponse(false, mobile_apis::Result::IN_USE, "");
+      return false;
+    }
+    case AcquireResult::ASK_DRIVER: {
+      SetResourceState(module_type, ResourceState::BUSY);
+      SendGetUserConsent(module_type);
```

Comment:
Don't know got how to allow 20 seconds timeout for the request in this special case (10s default + 10s for waiting pop up)